### DISCUSSION TITLE
subscribeRepos: "caught up" #info message?

### DISCUSSION THREAD

**Discussion by snarfed** [bluesky-social/atproto#1516] [Protocol (atproto)]
Right now, if you `subscribeRepos` with a cursor, the server will backfill historical commits until it catches up to the present, and then serves commits as they happen. It'd be useful in some cases to know when you've "caught up" to the present, eg if you want to backfill but not stay subscribed in realtime. `#info` seems like the perfect vehicle for this, maybe via a new `CaughtUp` value for `name`?

As an example, Nostr's query protocol is somewhat similar to `subscribeRepos` in that it combines historical querying with realtime subscription over websocket. [It signals "caught up" with the `EOSE` (end of stored events) message.](https://github.com/nostr-protocol/nips/blob/master/01.md#from-relay-to-client-sending-events-and-notices)

**Comment by bnewbold** [comment]
It is an interesting idea. I'd want to have more experience and confidence building firehose producers before committing to this, as it might have tricky concurrency requirements in some implementations.

Other approaches would be passing a query param when starting the subscription. Or getting the current offset cursor (one way or another), then the client can just drop the connection when it gets to the end.

**Reply by snarfed** [reply]
Sure! Those could work for this specific use case. There may be other use cases that want both backfill and realtime, and want to handle them differently, but I don't personally have those use cases yet, so no need to design for hypotheticals.
